### PR TITLE
Migrate data from legacy app

### LINKED ISSUE #1070: Read legacy app's database, and migrate data
When we launch this app, it will go out as the new version of the Zulip mobile app on the Play Store and App Store. That means that users who already have the Zulip app installed, and upgrade, will get the new app replacing the legacy app.

At that point, if we don't do any special work to make something else happen, the user will find that the app has effectively forgotten their login to whatever Zulip server or servers they were using, as well as any of the local settings they had in the app. That's because the new app stores its data in a different place and format from the old app, and simply won't see the old data.

It'd be nice to instead make it so that the logins and settings carry over from the legacy app.

### Implementation

This will involve looking carefully through the details of how the legacy app stores its data, in order to write logic in the new app to consume it. Here's an outline; details are to be found in the legacy app's code, and in the code of its dependencies.

1. We’ll need to read the SQLite database. This is straightforward — we're using SQLite anyway, for the new app's own database.
2. Then, on Android, we’ll decompress the data.
   * ​For this, we can always just take the relevant chunk of Kotlin code (which we wrote) and wrap it in a Flutter plugin instead of in a React Native native module.
   * ​Alternatively and perhaps cleaner, we can also transcribe that logic into Dart instead.
3. Then we’ll deserialize the JSON.
4. ​We’ll drop most of the data — the stuff that’s fundamentally a cache for data that lives on the server — just as we do for most migrations within the RN app.
5. ​We’ll apply any needed migrations from the history of the RN app. We can basically just transcribe the existing code to Dart.
6. ​Finally, we have data in the latest schema from the RN app. We’ll write it out in the new app’s own schema, and load it from there.

Then to be useful, this logic needs to be correct the very first time it runs on any given user's device. It therefore needs to be thoroughly tested.

Here's some optional bonus extra steps. We'll skip them, because they'd only be relevant for devices upgrading directly from a zulip-mobile release older than v27.183 — which was dated 2022-04-12, over two and a half years ago now. (I originally wrote these down when this new Flutter-based app was just a proposal, partly in order to make the point that we could do the migration very thoroughly. Copying them here now mostly for readers' entertainment.)

<details><summary>Steps for unnecessarily comprehensive migration</summary>

* For step 1 above, "read the SQLite database", add the following sub-steps:
  * ​The SQLite database is what we’ll find for any install that’s run a version released since v27.183, back in 2022-04. The relevant PR was: https://github.com/zulip/zulip-mobile/pull/5309
  * ​For older app versions, on Android, the data is simply in a SQLite database at a different file path.
​  * ​For older app versions, on iOS, the data is in a more ad-hoc format, because that’s what RN chose to do on iOS for its “AsyncStorage” feature. Still straightforward, though more annoying:
    * ​Specifically, the whole thing is JSON-serialized to a file somewhere; except that where the values are bigger than some threshold (a few kB?), they get written separately to their own files.
    * ​One option is always to just take the relevant chunk of Objective-C code — it’s a few hundred lines — and vendor it, wrapped in a Flutter plugin instead of in a React Native native module.
​    * Alternatively and perhaps cleaner, we can transcribe that logic into Dart.

</details> 

### Scope

The data that's relevant here — the data that the legacy app stores, and that we won't discard in step 4 above — is listed here (from `src/boot/store.js` in the legacy app):
```js
export const storeKeys: $ReadOnlyArray<$Keys<GlobalState>> = [
  'migrations', 'accounts', 'drafts', 'outbox', 'settings',
];
```

The `migrations` row is needed for step 5 above. Other than that, `accounts` is the most important; it corresponds to our `Accounts` table, or `globalStore.accounts`.

The other three correspond to features we may or may not have implemented at launch time:
* `settings`: #97 
* `outbox`: #133 
* `drafts`: (We don't have an issue for this in the tracker. Not a priority.)

### PR DESCRIPTION
**Pull request by gnprice**
Fixes #1070.

This is a draft because I'd like to write at least some tests before merging. (Unlike some of the other PRs we've merged in the past week with tests left as a follow-up, the job this is doing is especially sensitive to edge cases.)

But otherwise I believe this is ready. I've tested it manually on both Android and iOS, and the data gets carried along as intended. So in parallel with me writing those tests, I'd like to have additional eyes reading the code, to try to spot things I might have missed.

### CODE REVIEW

File: lib/model/legacy_app_data.dart [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+/// This and its subclasses correspond to portions of src/storage/replaceRevive.js .
+///
+/// (The rest of the conversions in that file are for types that don't appear
+/// in the portions of the legacy app's state we care about.)
+sealed class _LegacyAppJsonConverter<T> extends JsonConverter<T, Map<String, dynamic>> {
+  const _LegacyAppJsonConverter();
+
+  static const List<_LegacyAppJsonConverter<Object?>> converters = [
+    _LegacyAppUrlJsonConverter(),
```
Comment (PIG208) [inline lib/model/legacy_app_data.dart]: I can see why `_LegacyAppZulipVersionJsonConverter` got left out here, since that converter has `T` as `String`. A comment explaining that this is intentional might be helpful.

(Or perhaps just inline the `_LegacyAppUrlJsonConverter` annotation when needed?)

**Comment by PIG208** [comment]
I probably haven't set this up properly. On my Android debug build, `LegacyAppDatabase._filename` returns `/data/user/0/com.zulip.flutter/files/SQLite/zulip.db`. That path, with "com.zulip.flutter" replaced with "com.zulipmobile", does not exist; instead, I got `Android/data/com.zulipmobile/files` (which I think I don't have access to).

File: lib/model/legacy_app_data.dart [hunk 2]
```diff
@@ -8,13 +8,83 @@ library;
...
+    return;
+  }
+
+  assert(debugLog("Found settings: ${legacyData.settings?.toJson()}"));
+  final settings = legacyData.settings;
+  if (settings != null) {
+    await db.update(db.globalSettings).write(GlobalSettingsCompanion(
+      // TODO(#1139) apply settings.language
+      themeSetting: switch (settings.theme) {
+        LegacyAppThemeSetting.default_ => drift.Value.absent(),
```
Comment (PIG208) [inline lib/model/legacy_app_data.dart:46]: I think the "default" in the legacy app [is light mode](https://github.com/zulip/zulip-mobile/blob/eb8505c4a0ccfb7bdd817ca9022c11ce57ab69d8/src/settings/settingsSelectors.js#L5-L6).

There was a migration to rename "default" to "light" (https://github.com/zulip/zulip-mobile/commit/43acc4199919d5ea8ece1d00ac6e1243b4073e67). Oh, but that was ultimately for https://github.com/zulip/zulip-mobile/issues/5533 (Automatically follow user's OS-level dark/light preference). So I think, after all, this completes:

> On existing installs where the user has stuck with the default of light mode, we'll give them the new default.

**Comment by gnprice** [comment]
Thanks for the review!

For investigating the file paths, I started a chat thread: [#mobile-team > migrating legacy data @ 💬](https://chat.zulip.org/#narrow/channel/243-mobile-team/topic/migrating.20legacy.20data/near/2194721). Let's continue there for debugging.

(I'll also reply on the other subthreads above, but that one is the most critical since it could affect whether the migration works at all for a given user.)

File: lib/model/legacy_app_data.dart (hunk 2 above)
Comment (gnprice) [reply lib/model/legacy_app_data.dart:46]: Yeah, this is a good question to consider explicitly.

The legacy app has just two values: always light and always dark. In the app's data store, and so in the database, these are called "default" and "night" respectively (that's the `ThemeSetting` enum cited in a comment here). And the default is "default" (i.e. thankfully the name isn't misleading on that point).

The new app has values `light`, `dark`, and null. Null means to use the system-wide setting, and is the default.

So if the value in the legacy app's database is "night", clearly we should translate that to `dark`. The user explicitly chose a dark theme.

If the value in the legacy app's database is "default", though, it's less clear. That's the value it will have if the user didn't make any choice.

So my thinking is that if the value is "default", then we should translate that to the new default, which is to use the system-wide setting. If the system-wide setting is on a daily cycle, then the user will almost certainly experience that as an upgrade: at last Zulip follows the daily cycle the rest of the device follows. If the system-wide setting is light, there'll be no change. And if the system-wide setting is dark, but the user really does prefer Zulip specifically to be in a light theme… well, it's easy for them to change it.

File: lib/model/legacy_app_data.dart (hunk 1 above)
Comment (gnprice) [reply lib/model/legacy_app_data.dart]: Yeah, this could be clearer. I started with this one; then got to where `ZulipVersion` is needed and the same approach doesn't apply.

Given that there ended up being just one place `_LegacyAppUrlJsonConverter` comes into effect (for `LegacyAppAccount.realm`), and that this set of types is basically complete now and isn't going to expand in the future, I think I'll just apply the converter explicitly in that one spot and drop this list.

File: lib/model/legacy_app_data.dart (hunk 2 above)
Comment (gnprice) [reply lib/model/legacy_app_data.dart:46]: Ah, and GitHub was showing me a stale page that didn't have your edit. Yeah, exactly, this effectively completes that issue https://github.com/zulip/zulip-mobile/issues/5533 .

File: lib/model/legacy_app_data.dart [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+      // (See docs and comments on these properties in src/types.js .)
+      // Drop the account.
+      assert(debugLog("    (account ignored because missing metadata)"));
+      continue;
+    }
+    await db.createAccount(AccountsCompanion.insert(
+      realmUrl: account.realm,
+      userId: account.userId!,
+      email: account.email,
+      apiKey: account.apiKey,
```
Comment (PIG208) [inline lib/model/legacy_app_data.dart:96]: Hmm, what happens if `apiKey` is an empty string ([possible for logged-out users](https://github.com/zulip/zulip-mobile/blob/eb8505c4a0ccfb7bdd817ca9022c11ce57ab69d8/src/api/transportTypes.js#L21-L27))?

I think the converted `Account` will show up on the choose-account page, and will get handled by the code responsible for invalid API keys. The Flutter app doesn't remember logged-out accounts right now, similar to how language setting is not yet supported. Perhaps this can be a TODO?

Comment (gnprice) [reply lib/model/legacy_app_data.dart:96]: Hmm, yeah, good find.

The new app doesn't have a concept of an account that's in the list and yet logged out.

I'm not sure it was all that useful in the legacy app either — it's basically a relic of the fact that, when you were a screen or two into the login flow, the legacy app didn't originally have a better place to keep that information (with the realm URL, and another screen later the email and password, that you'd already entered) than in its global persistent list of accounts. We eventually fixed that, making that information part of the relevant pages' local state instead, and the new app does the same thing.

Since the new app doesn't have this concept, I'll have it drop such records if they appear.

**Comment by gnprice** [comment]
Revision pushed. Thanks @PIG208 for the review!

File: lib/model/legacy_app_data.dart [hunk 4]
```diff
@@ -8,13 +8,99 @@ library;
...
+      zulipVersion: account.zulipVersion!,
+      // no zulipMergeBase; legacy app didn't record it
+      zulipFeatureLevel: account.zulipFeatureLevel!,
+      ackedPushToken: drift.Value(account.ackedPushToken),
```
Comment (chrisbobbe) [inline lib/model/legacy_app_data.dart]: This will be the first time we store a non-null `ackedPushToken` in the new app, since we haven't done #322 yet. We don't update it when the server registers a new push token, so it might become stale…currently I think such a stale token could only flow into an unregister-token request, which we do on logout.

Comment (gnprice) [reply lib/model/legacy_app_data.dart]: Hmm interesting, good catch.

I think I'll drop this value, then — it doesn't really do any good (since until #322 it doesn't cause this app to stop repeatedly sending the token to the server), and at that one spot where it would get used I think it would be more likely to make the unregister-token request not work (by making it use a stale token) than to make it work. (Both of which are unlikely — I believe these tokens rarely change in practice.)

When we eventually do #322, that will cause the app to promptly store the current value, and there won't be any benefit to remembering the value the legacy app had stored in the past.

**Comment by gnprice** [comment]
(Pushed a revision with that change.)

**Comment by gnprice** [comment]
OK, at this point I'm not going to get to writing any significant automated tests for this before the launch. So I'll file a followup issue for those, and go ahead and merge after tweaking the commit messages.

(I also just pushed a revision expanding one comment, with details on when the legacy app's Account records started having a `userId`.)

Effectively we'll just rely on
* the manual testing I did before sending this
* the scrutiny of the code by @PIG208 and @chrisbobbe above (plus my own self-review before that)
* and the manual testing we [got today](https://chat.zulip.org/#narrow/channel/243-mobile-team/topic/last.20code.20changes.20for.20launch/near/2194853) from alpha users.

Speaking of that alpha testing, I have a couple more edits to harden the logic that I want to include for launch. I'll send those as a separate follow-up PR, though — that way the merged version of this in main can be the same logic that went into today's v30.0.257 beta, and the difference between the logic in v30.0.257 and the next build (for launch) can be expressed as a discrete commit of its own. That's helpful for looking at the history in Git to understand the history of the releases.

**Comment by gnprice** [comment]
And merged. Thanks again @PIG208 and @chrisbobbe for the reviews!
